File: 1905. Count Sub Islands.java
class Solution {
    private static final int[][] directions = {{1,0},{-1,0},{0,1},{0,-1}};
    public int countSubIslands(int[][] grid1, int[][] grid2) {
        int count = 0;
        for (int i = 0; i < grid2.length; i++) {
            for (int j = 0; j < grid2[i].length; j++) {
                if (grid2[i][j] == 1) {
                    if (floodFill(grid2, grid1, i, j, true)) {
                        count++;
                    }
                }
            }
        }
        return count;
    }
    private boolean floodFill(int[][] grid2, int[][] grid1, int i, int j, boolean sub) {
        if (i < 0 || i >= grid2.length || j < 0 || j >= grid2[i].length || grid2[i][j] == 0) {
            return sub;
        }
        if (grid1[i][j] == 0) sub = false;
        grid2[i][j] = 0;
        for (int[] d : directions) {
            sub = floodFill(grid2, grid1, i+d[0], j+d[1], sub) && sub;
        }
        return sub;
    }
}

File: go/1905. Count Sub Islands.go
func countSubIslands(grid1 [][]int, grid2 [][]int) int {
    n, m := len(grid2), len(grid2[0])
    answer := 0
    for i := 0; i < n; i++ {
        for j := 0; j < m; j++ {
            if grid2[i][j] == 1 {
                if floodFill(grid2, grid1, i, j, n, m, true) {
                    answer++
                }
            }
        }
    }
    return answer
}

func floodFill(grid2, grid1 [][]int, x, y, n, m int, answer bool) bool {
    if x >= n || x < 0 || y >= m || y < 0 { return answer }
    if grid2[x][y] == 0 { return answer }
    grid2[x][y] = 0
    if grid1[x][y] == 0 { answer = false }
    answer = floodFill(grid2, grid1, x+1, y, n, m, answer)
    answer = floodFill(grid2, grid1, x-1, y, n, m, answer)
    answer = floodFill(grid2, grid1, x, y+1, n, m, answer)
    answer = floodFill(grid2, grid1, x, y-1, n, m, answer)
    return answer
}
